prover/src/lib.rs

```diff
@@ -131,13 +130,16 @@ pub struct ProverInput {
 }
 
 impl ProverInput {
+    /// Prove all the blocks in the input.
+    /// Return the list of block numbers that are proved and if the proof data
+    /// is not saved to disk, return the generated block proofs as well.
     pub async fn prove(
         self,
         runtime: &Runtime,
         previous_proof: Option<GeneratedBlockProof>,
         save_inputs_on_error: bool,
         proof_output_dir: Option<PathBuf>,
-    ) -> Result<Vec<BlockNumber>> {
+    ) -> Result<Vec<(BlockNumber, Option<GeneratedBlockProof>)>> {
```

Comment:
I'm not sure I see the need of having each returned proof as an `Option`. Either the process aborts early or returns a valid proof, and we don't seem to be using the fact that this is wrapped in an `Option` anywhere else. Is it because you don't want `stdio` mode to write to `stdout` proofs that would have already been written to disk?

Reply:
The same prove function is used in both `jerigon` and `stdio` mode. In jerigon we do not return proof, in `stdio` it is much easier to formate complete proof output if it is not partially output on stdio for every generated block proof

Reply:
But if I understand correctly, writing to disk will set the `Option` to `None` for the given proof, which will then be filtered away before calling the final writer of `stdio` method, effectively not writing anything anymore to `stdout` unlike what we currently do, no?

Reply:
No, we return list of proved block heights anyway, so `stdio` method will fully output proof (which passed CI tests with verification show).

Reply:
We use `proof_output_dir` to configure if we want proofs written to disk. If it is none, proofs will be returned from `prove` function, in both `jerigon` and `stdio` mode

Reply:
Yes, but if `proof_output_dir` is `Some`, we write the proof to disk, then return `Ok((block_number, None))`, which will then be filtered away when calling
```rust
let proofs: Vec<GeneratedBlockProof> = proved_blocks
        .into_iter()
        .filter_map(|(_, proof)| proof)
        .collect();
```
in `leader/src/stdio.rs`, effectively having `proofs` be empty (hence not writing anything to `stdout`. This is what I am saying is different from the current behavior.

Reply:
In `stdio` mode we always set `proof_output_dir` to `None`:
```
let proved_blocks = prover_input
        .prove(&runtime, previous, save_inputs_on_error, None)
        .await;
```

Reply:
Ah ok that's what I was missing. All good then.